File: src/ai/ranks.ts
/** 난이도 10단계 → KataGo visits / 내장 휴리스틱 */

export type RankId =
  | 'lv1'
  | 'lv2'
  | 'lv3'
  | 'lv4'
  | 'lv5'
  | 'lv6'
  | 'lv7'
  | 'lv8'
  | 'lv9'
  | 'lv10'

export interface RankOption {
  id: RankId
  /** 1–10 */
  level: number
  labelKo: string
  labelEn: string
  /** KataGo maxVisits (대략) — 저레벨은 엔진 미사용 */
  visits: number
  /** 한 수 최대 초 (GPU에서도 체감 상한) */
  maxTimeSec: number
  /** 내장 AI: 합법적 수 중 무작위 비율 (1=완전랜덤, 0=캡처/중앙 선호) */
  randomness: number
  /**
   * KataGo 후보 선택 온도. 높을수록 약한 수도 자주 둠.
   * 내장 AI만 쓰는 레벨에서는 무시.
   */
  moveTemperature: number
}

export const RANKS: RankOption[] = [
  { id: 'lv1', level: 1, labelKo: '1 · 입문', labelEn: '1 · Beginner', visits: 1, maxTimeSec: 0.1, randomness: 1.0, moveTemperature: 5 },
  { id: 'lv2', level: 2, labelKo: '2 · 왕초보', labelEn: '2 · Novice', visits: 1, maxTimeSec: 0.15, randomness: 0.97, moveTemperature: 4 },
  { id: 'lv3', level: 3, labelKo: '3 · 초급', labelEn: '3 · Easy', visits: 2, maxTimeSec: 0.25, randomness: 0.9, moveTemperature: 3 },
  { id: 'lv4', level: 4, labelKo: '4 · 초급+', labelEn: '4 · Easy+', visits: 4, maxTimeSec: 0.4, randomness: 0.8, moveTemperature: 2.2 },
  { id: 'lv5', level: 5, labelKo: '5 · 중급', labelEn: '5 · Medium', visits: 12, maxTimeSec: 0.7, randomness: 0.45, moveTemperature: 1.4 },
  { id: 'lv6', level: 6, labelKo: '6 · 중급+', labelEn: '6 · Medium+', visits: 24, maxTimeSec: 1.0, randomness: 0.3, moveTemperature: 1.0 },
  { id: 'lv7', level: 7, labelKo: '7 · 상급', labelEn: '7 · Hard', visits: 48, maxTimeSec: 1.3, randomness: 0.18, moveTemperature: 0.6 },
  { id: 'lv8', level: 8, labelKo: '8 · 상급+', labelEn: '8 · Hard+', visits: 100, maxTimeSec: 1.8, randomness: 0.1, moveTemperature: 0.35 },
  { id: 'lv9', level: 9, labelKo: '9 · 유단', labelEn: '9 · Dan', visits: 180, maxTimeSec: 2.4, randomness: 0.05, moveTemperature: 0.2 },
  { id: 'lv10', level: 10, labelKo: '10 · 고수', labelEn: '10 · Expert', visits: 320, maxTimeSec: 3.2, randomness: 0.02, moveTemperature: 0.1 },
]

export const DEFAULT_RANK: RankId = 'lv3'

/** KataGo 신경망은 visits=1이어도 강함 → 입문~초급+는 내장 랜덤 AI만 사용 */
export function usesKataGoEngine(rankId: RankId): boolean {
  return getRank(rankId).level >= 5
}

export function getRank(id: RankId): RankOption {
  return RANKS.find((r) => r.id === id) ?? RANKS[2]
}

export function rankLabel(id: RankId, lang: string): string {
  const r = getRank(id)
  return lang === 'ko' ? r.labelKo : r.labelEn
}

File: app/lib/domain/ai/ranks.dart
// lib/domain/ai/ranks.dart — 난이도 10단계
//
// KataGo 신경망은 visits=1 이어도 사람보다 강하다. 그래서 저레벨은
// 엔진을 아예 안 쓰고 내장 휴리스틱의 무작위 비율로만 약하게 만든다.

class RankOption {
  const RankOption({
    required this.id,
    required this.level,
    required this.visits,
    required this.maxTime,
    required this.randomness,
    required this.moveTemperature,
  });

  final String id;

  /// 1~10
  final int level;

  /// KataGo maxVisits (대략)
  final int visits;

  /// 한 수 상한 — GPU 에서도 체감되는 값이라 상한을 둔다
  final Duration maxTime;

  /// 합법 수 중 무작위로 고르는 비율 (1.0 = 완전 랜덤)
  final double randomness;

  /// KataGo 후보 선택 온도. 내장 AI 전용 레벨에서는 무시된다.
  final double moveTemperature;

  /// i18n 키 — 라벨을 코드에 박지 않는다 (하드코딩 문자열 0 규칙)
  String get labelKey => 'rank_$id';
}

const List<RankOption> kRanks = <RankOption>[
  RankOption(id: 'lv1', level: 1, visits: 1, maxTime: Duration(milliseconds: 100), randomness: 1.0, moveTemperature: 5),
  RankOption(id: 'lv2', level: 2, visits: 1, maxTime: Duration(milliseconds: 150), randomness: 0.97, moveTemperature: 4),
  RankOption(id: 'lv3', level: 3, visits: 2, maxTime: Duration(milliseconds: 250), randomness: 0.9, moveTemperature: 3),
  RankOption(id: 'lv4', level: 4, visits: 4, maxTime: Duration(milliseconds: 400), randomness: 0.8, moveTemperature: 2.2),
  RankOption(id: 'lv5', level: 5, visits: 12, maxTime: Duration(milliseconds: 700), randomness: 0.45, moveTemperature: 1.4),
  RankOption(id: 'lv6', level: 6, visits: 24, maxTime: Duration(milliseconds: 1000), randomness: 0.3, moveTemperature: 1.0),
  RankOption(id: 'lv7', level: 7, visits: 48, maxTime: Duration(milliseconds: 1300), randomness: 0.18, moveTemperature: 0.6),
  RankOption(id: 'lv8', level: 8, visits: 100, maxTime: Duration(milliseconds: 1800), randomness: 0.1, moveTemperature: 0.35),
  RankOption(id: 'lv9', level: 9, visits: 180, maxTime: Duration(milliseconds: 2400), randomness: 0.05, moveTemperature: 0.2),
  RankOption(id: 'lv10', level: 10, visits: 320, maxTime: Duration(milliseconds: 3200), randomness: 0.02, moveTemperature: 0.1),
];

const String kDefaultRank = 'lv3';

RankOption getRank(String id) =>
    kRanks.firstWhere((RankOption r) => r.id == id,
        orElse: () => kRanks[2]); // 못 찾으면 기본 난이도

/// 입문~초급+ 는 내장 AI 만 쓴다 — 신경망은 최저 설정에서도 너무 강하다
bool usesKataGoEngine(String id) => getRank(id).level >= 5;
